discovery-provider/src/queries/get_notifications.py

```diff
@@ -0,0 +1,221 @@
+import logging
+from collections import defaultdict
+from datetime import datetime
+from typing import Dict, List, Optional, TypedDict, Union
+
+from sqlalchemy import bindparam, text
+from sqlalchemy.orm.session import Session
+
+logger = logging.getLogger(__name__)
+
+
+class GetNotificationArgs(TypedDict):
+    user_id: int
+    timestamp: Optional[int]
+    limit: Optional[int]
+
+
+notification_groups_sql = text(
+    """
+--- Create Intervals of user seen
+WITH user_seen as (
+  SELECT
+    LAG(seen_at, 1, now()::timestamp) OVER ( ORDER BY seen_at desc ) AS seen_at,
+    seen_at as prev_seen_at
+  FROM
+    notification_seen
+---  Not sure if we should add the following - redundant, but could be faster/slower
+---  WHERE
+---    (:timestamp_offset is NULL or seen_at < :timestamp_offset)
+  WHERE
+    user_id = :user_id
+---  Not sure if we should add the following - redundant, but could be faster/slower
+---  ORDER BY
+---    seen_at desc
+)
+SELECT
+    n.group_id as group_id,
+    array_agg(n.id),
+    CASE 
+      WHEN user_seen.seen_at is not NULL THEN now()::timestamp != user_seen.seen_at
+      ELSE EXISTS(SELECT 1 from notification_seen where user_id = :user_id)
+    END as is_seen,
+    CASE 
+      WHEN user_seen.seen_at is not NULL THEN user_seen.seen_at
+      ELSE (
+        SELECT seen_at from notification_seen
+        WHERE user_id = :user_id
+        ORDER BY seen_at ASC
+        limit 1
+      )
+    END as seen_at,
+    user_seen.prev_seen_at,
+    count(n.group_id)
+FROM
+    notification n
+LEFT JOIN user_seen on
+  user_seen.seen_at >= n.timestamp and user_seen.prev_seen_at < n.timestamp
+WHERE
+  :user_id = ANY(n.user_ids) AND
+  (:timestamp_offset is NULL or n.timestamp <= :timestamp_offset) AND
```

Comment:
hmmm, well because we are paginating backwards and because the timestamps can easily be the same - given that they are set from block number - the the pagination can occur between blocknumber which is why there's a secondary sort for group_id that does use `<`
lmk if that makes sense